[flang][Driver] Preliminary support for -ftime-report

tarunprabhu opened the pull request:
The behavior is not entirely consistent with that of clang for the moment since detailed timing information on the LLVM IR optimization and code generation passes is not provided. The -ftime-report= option is also not enabled since that is only relevant for information about the LLVM IR passes. However, some code to handle that option has been included, to make it easier to support the option when the issues blocking it are resolved. A FortranSupport library has been created that is intended to mirror the LLVM and MLIR support libraries.

-----------------------------
Comments for reviewers: 

Support for detailed information about the LLVM passes has been commented out because it results in a duplicate registration of the `-sort-times` LLVM command-line option. I have no idea why this is happening. The only way I was able to resolve it was by removing the call `InitLLVM()`. `check-flang` still passed after the call was removed, but I only tested on Linux on x86. If anyone has suggestions on a better way to resolve this, I am happy to take them on board. 

FWIW, `clang` does not seem to call `InitLLVM()`.

That being said, the lines that would enable this support are the ones below (`clang` does precisely this as well):
```
    // llvm::TimePassesIsEnabled = true;
    // llvm::TimePassesPerRun = invoc.getTimeLLVMPassesPerRun();
```
This essentially sets some global variables in LLVM (which are also set when the `-ftime-passes` LLVM command-line option is used). In [a different PR](https://github.com/llvm/llvm-project/pull/106141), @banach-space noted that such approaches ought to be avoided and I tend to agree. One way to resolve the issue would be to provide some hooks in LLVM that will allow the timers to be initialized independently of the command line options. I can put up a draft PR for this is anyone is interested.

Commit:
[flang][Driver] Preliminary support for -ftime-report

The behavior is not entirely consistent with that of clang for the moment
since detailed timing information on the LLVM IR optimization and code
generation passes is not provided. The -ftime-report= option is also not enabled
since that is only relevant for information about the LLVM IR passes. However,
some code to handle that option has been included, to make it easier to support
the option when the issues blocking it are resolved. A FortranSupport library
has been created that is intended to mirror the LLVM and MLIR support libraries.

Commit:
Run clang-format

tblah commented on flang/test/Driver/time-report-eq.f90:
```diff
@@ -0,0 +1,18 @@
+! timing information on the LLVM IR optimization and code generation passes.
+! When that is supported, these can be re-enabled.
+!
+! XFAIL: *
```
Maybe we should save this test for a PR enabling -ftime-report=?

tblah reviewed (commented):
Thanks for working on this

tarunprabhu commented on flang/include/flang/Frontend/CompilerInstance.h:
```diff
@@ -254,6 +276,43 @@ class CompilerInstance {
+  /// {
+  /// @name Timing
+  /// @{
+  bool isTimingEnabled() { return timingMgr.isEnabled(); }
```
Good point. I will remove it.

tarunprabhu commented on flang/include/flang/Frontend/CompilerInvocation.h:
```diff
@@ -143,6 +144,14 @@ class CompilerInvocation : public CompilerInvocationBase {
+
+  /// Whether to report the timing of each run of an LLVM pass. Set when
+  /// -ftime-report=per-pass-run is enabled.
+  bool timeLLVMPassesPerRun;
```
That is reasonable. I will remove some of the related code and add TODO's in their place.

tarunprabhu replied:
I ended up not adding any TODOs. It didn't seem necessary.

tarunprabhu commented on flang/lib/CMakeLists.txt:
```diff
@@ -4,6 +4,7 @@ add_subdirectory(Decimal)
 add_subdirectory(Lower)
 add_subdirectory(Parser)
 add_subdirectory(Semantics)
+add_subdirectory(Support)
```
Mainly to match the layout in both `llvm` and `mlir`. In principle, it could be used in other tools, both in-tree and out-of-tree. It feels odd to have to link against the entire frontend just to get a timer.

tarunprabhu commented on flang/lib/Frontend/CompilerInstance.cpp:
```diff
@@ -176,6 +205,26 @@ bool CompilerInstance::executeAction(FrontendAction &act) {
+    timingMgr.setOutput(
+        Fortran::support::createTimingFormatterText(mlir::thread_safe_nulls()));
+
+    // This is deliberately done in "reverse" order and does not match the
+    // behavior of clang.
```
The printing of timings in the case of clang is done when the timer handler's destructors are run. I don't think they are guaranteed to run in a specific order. In this case, we are guaranteeing that the timings will be printed in "reverse", starting from code generation, then LLVM IR optimizations and then the MLIR and parsing. 

I will edit the comment.

Commit:
Address reviewer comments. Remove code specific to -ftime-report= since that
is not supported right now anyway.

Add checks that will only print out timing streams if there is anything in them
to avoid empty lines.

tblah reviewed (commented):
Thanks for the quick updates. I don't have any idea off the top of my head how to fix the LLVM timings. Do any other reviewers have thoughts/experience in this area?

macurtis-amd commented on flang/lib/Frontend/CompilerInstance.cpp:
```diff
@@ -167,6 +170,31 @@ bool CompilerInstance::executeAction(FrontendAction &act) {
+    // about the runtime of the LLVM IR optimization and code generation passes.
+    // Once the root cause of this is determined, we should enable this to have
+    // behavior that is comparable to clang.
+    // llvm::TimePassesIsEnabled = true;
```
@tarunprabhu #121663 has been merged which fixes the assert you were seeing. You should be able to uncomment this line (and remove preceding comment).

macurtis-amd commented on flang/lib/Frontend/CompilerInvocation.cpp:
```diff
@@ -1358,6 +1358,12 @@ bool CompilerInvocation::createFromArgs(
   }
 
+  // Process the timing-related options.
+  if (const llvm::opt::Arg *a =
```
`a` is unused and produces a warning in my build.

macurtis-amd commented on flang/include/flang/Frontend/CompilerInvocation.h:
```diff
@@ -222,6 +227,9 @@ class CompilerInvocation : public CompilerInvocationBase {
     return defaultKinds;
   }
 
+  bool getEnableTimers() { return enableTimers; }
```
Nit: non-const getter is not needed.

MaskRay commented on flang/lib/Frontend/FrontendActions.cpp:
```diff
@@ -81,6 +81,16 @@
 
 using namespace Fortran::frontend;
 
+static constexpr llvm::StringLiteral timingIdParse = "Parse";
```
constexpr at namespace level implies static

MaskRay commented on flang/test/Driver/time-report.f90:
```diff
@@ -0,0 +1,22 @@
+! Check that -ftime-report flag is passed as-is to fc1. The value of the flag
+! is only checked there. This behavior intentionally mirrors that of clang.
+!
+! RUN: %flang -### -c -ftime-report %s 2>&1 | FileCheck %s --check-prefix=CHECK-DRIVER
```
do you need -mllvm -sort-timers=0 to prevent reordering of timers in the TimerGroup?

tarunprabhu commented:
Superseded by #122894
